txpipe/magnification.py

```diff
@@ -0,0 +1,204 @@
+from .base_stage import PipelineStage
+from .data_types import (
+    YamlFile,
+    HDFFile,
+    FitsFile,
+    PNGFile,
+)
+from .utils import LensNumberDensityStats, Splitter, rename_iterated
+from .binning import build_tomographic_classifier, apply_classifier
+import numpy as np
+import warnings
+
+
+class TXSSIMagnification(PipelineStage):
+    """
+    class for computing the magnification coefficients using SSI outputs
+    Following the methodology of https://arxiv.org/abs/2012.12825
+    and https://arxiv.org/abs/2209.09782
+    """
+
+    name = "TXSSIMagnification"
+
+    inputs = [
+        ("binned_lens_catalog_nomag", HDFFile),
+        ("binned_lens_catalog_mag", HDFFile),
+    ]
+
+    outputs = [
+        ("magnification_coefficients", HDFFile),
+        ("magnification_plot", PNGFile),
+    ]
+
+    config_options = {
+        "chunk_rows": 10000,
+        "applied_magnification": 1.02, #magnification applied to the "magnified" SSI catalog
+        #TO DO: add a way for "applied_magnification" to be determined from the SSI inputs directly
+        "n_patches":20,
+        "bootstrap_error":True,
+    }
+
+    def run(self):
+        """
+        Run the analysis for this stage.
+        """
+        from scipy.stats import bootstrap
+
+        # load the catalogs
+        nomag_cat = self.open_input("binned_lens_catalog_nomag")
+        mag_cat   = self.open_input("binned_lens_catalog_mag")
+
+        # get/estimate the magnification applied to the magnified catalog
+        # TO DO: could put this as optional config item in TXSSIIngest
+        # so that it is done automatically
+        mu = self.config["applied_magnification"]
+        deltak = (1. - 1./mu)/2.
+        
+        if self.config["bootstrap_error"]:
+            #split no-mag sample up into patches to be used in the bootstrap errors
+            cluster = self.calc_cluster_patches(nomag_cat)
+
+        # compute fractional change in number count for both samples
+        # in each lens bin
+        nbins = nomag_cat['lens/'].attrs['nbin_lens']
+        outfile = self.setup_output(nbins)
+        for ibin in range(nbins):
+            print(f'Computing magnification coefficient for bin {ibin+1}')
+
+            # single redshift bins should be low enough number 
+            # count that we can load the whole sample
+            w1 = nomag_cat[f'lens/bin_{ibin}/weight'][:]
+            w2 = mag_cat[f'lens/bin_{ibin}/weight'][:]
+
+            #compute mag coeff with the whole sample
+            csample = self.calc_frac_change(w1, w2)/deltak
+            self.write_output(outfile, csample, ibin)
+
+            if self.config["bootstrap_error"]:
+                print(f'Computing uncertainty with bootstrap')
+
+                #compute mag coeff with subsampling patches to get a boostrap error
+                patch1 = cluster.predict(
```

Comment:
Can you add a comment here explaining what's happening some more - e.g. what kind of object `cluster` is, and what we are `predict`ing.

Reply:
done